### PR TITLE
Add array_position function

### PR DESCRIPTION
**Pull request by haozhun**
Documents will be added when you agree with:
* the order of the parameters 
* the result when no match is found

### CODE REVIEW

**Comment by cberner** [comment]
@electrum could you assign back to me after you review the name? Alternatives we came up with are `array_index_of`

**Comment by electrum** [comment]
Let's call it `array_index`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return null;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: Add a description

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class ArrayFindFunction
+        extends ParametricScalar
+{
+    public static final ArrayFindFunction ARRAY_FIND = new ArrayFindFunction();
+    private static final Signature SIGNATURE = new Signature("array_find", ImmutableList.of(typeParameter("E")), "bigint", ImmutableList.of("E", "array<E>"), false, false);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: This should be `comparableTypeParameter`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: this should be `true`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Argument type to array_find unsupported");
+        }
+        return new FunctionInfo(
+                new Signature("array_find", parseTypeSignature(StandardTypes.BIGINT), type.getTypeSignature(), parameterizedTypeName("array", type.getTypeSignature())),
+                "Returns the cardinality (length) of the array",
+                false,
+                arrayFindMethodHandle.bindTo(type).bindTo(equalMethodHandle),
+                true,
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: use `isHidden()` and `isDeterministic()` to make sure these stay in sync and to make it easier to read

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        else if (type.getJavaType() == Slice.class) {
+            arrayFindMethodHandle = METHOD_HANDLE_SLICE;
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Argument type to array_find unsupported");
+        }
+        return new FunctionInfo(
+                new Signature("array_find", parseTypeSignature(StandardTypes.BIGINT), type.getTypeSignature(), parameterizedTypeName("array", type.getTypeSignature())),
+                "Returns the cardinality (length) of the array",
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: wrong description

**Comment by cberner** [comment]
I think the arguments should be in the opposite order to match `contains()`. @electrum thoughts?

**Comment by cberner** [comment]
Back to you for comment

**Comment by haozhun** [comment]
Fixed the comments and added documentation

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 6]
```diff
@@ -29,6 +29,10 @@ Array Functions
 
     Returns an array of the elements in the intersection of ``x`` and ``y``, without duplicates.
 
+.. function:: array_position(array, element) -> bigint
+
+    Returns the first occurrence of the ``element`` in ``array`` (or 0 if not found).
```
Comment (cberner) [inline presto-docs/src/main/sphinx/functions/array.rst]: should say "Returns the position of the first..."

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayPositionFunction.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns the first occurrence of the given value in array (or 0 if not found)";
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayPositionFunction.java]: "the position of the first..."

**Comment by cberner** [comment]
Couple minor comments. Looks good

**Pull request merged**